.github/workflows/test_functional.yml

```diff
@@ -147,17 +148,13 @@ jobs:
               -j "${NPROC}" \
               --state=save \
               ${{ matrix.test-base }} \
-              || (touch tests-did-not-time-out && false)
+              || (echo "::set-output name=finished::true" && false)
 
       - name: Re-run failed tests
         timeout-minutes: 10
-        if: steps.test.outcome == 'failure'
+        if: steps.test.outcome == 'failure' && steps.test.outputs.finished == 'true'
         run: |
-          if [[ ! -f tests-did-not-time-out ]]; then
-            echo "the preious test run timed out -> not re-running tests"
-            # this false causes the workflow to fail
-            false
-          fi
+          # re-run failed tests providing that they didn't time out first time
           # TODO: make the tests deterministic so we don't need to do this
```

Comment:
Not sure if the `continue-on-error` for the step `test` will cause the overall workflow to be marked as passed if `test` times out. It might be safer to keep the `if` inside the bash script, i.e.
```
if [[ ! ${{ steps.test.outputs.finished == 'true' } ]]; then
  echo "the preious test run timed out -> not re-running tests"
...
```

Reply:
Derp, that was half finished - 4910e40